Translation of Forms page (#100)
reactjs/pt-BR.reactjs.org: content/docs/forms.md

  - "docs/forms-zh-CN.html"
---

> Os elementos de formulário HTML funcionam de maneira um pouco diferente de outros elementos DOM no React, porque os elementos de formulário mantêm naturalmente algum estado interno. Por exemplo, este formulário em HTML simples aceita um único nome:

Reviewer: ```suggestion
Os elementos de formulário HTML funcionam de maneira um pouco diferente de outros elementos DOM no React, porque os elementos de formulário mantêm naturalmente algum estado interno. Por exemplo, este simples formulário em HTML  aceita um único nome:
```

```html
<form>
  <label>
    Nome:
    <input type="text" name="name" />
  </label>
  <input type="submit" value="Enviar" />
</form>
```

Esse formulário tem o comportamento padrão do HTML de navegar para uma nova página quando o usuário enviar o formulário. Se você quer esse comportamento no React, ele funciona. Mas na maioria dos casos, é conveniente ter uma função Javascript que manipula o envio de um formulário e tem acesso aos dados que o usuário digitou nos inputs. O modo padrão de fazer isso é com uma técnica chamada "componentes controlados".

## Componentes Controlados {#controlled-components}

Em HTML, elementos de formulário como `<input>`, `<textarea>` e `<select>` normalmente mantêm seu próprio estado e o atualiza baseado na entrada do usuário. Em React, o estado mutável é normalmente mantido na propriedade state dos componentes e atualizado apenas com [`setState()`](/docs/react-component.html#setstate).

> Podemos combinar os dois fazendo o estado React ser a "única fonte da verdade". Em seguida, o componente React que renderiza um formulário também controla o que acontece nesse formulário nas entradas subsequentes do usuário. Um input cujo o valor é controlado pelo React dessa maneira é chamado de "componente controlado".

Author: Eu ia perguntar mesmo na descrição da PR qual a melhor forma de traduzir "input". acho que dependendo do contexto "input do usuário" fica melhor. Irei corrigir.

  }

  handleSubmit(event) {
    alert('Um nome foi enviado: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Nome:
          <input type="text" value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Enviar" />
      </form>
    );
  }
}
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/VmmPgp?editors=0010)

Como o atributo `value` é definido no nosso `<input type="text">`, o valor exibido sempre será o mesmo de `this.state.value`, fazendo com que o estado do React seja a fonte da verdade. Como o `handleChange` é executado a cada tecla pressionada para atualizar o estado do React, o valor exibido será atualizado conforme o usuário digita.

> Com um componente controlado, cada mutação de estado terá uma função de manipulação associada. Isso faz com que seja simples modificar ou validar a entrada do usuário. Por exemplo, se quiséssemos impor que os nomes sejam escritos com todas as letras maiúsculas, poderíamos escrever `handleChange` como:

Reviewer 2: ```suggestion
Com um componente controlado, cada mutação de estado terá uma função de manipulação (handler function) associada. Isso faz com que seja simples modificar ou validar a entrada do usuário. Por exemplo, se quiséssemos impor que os nomes sejam escritos com todas as letras maiúsculas, poderíamos escrever `handleChange` como:
```

```javascript{2}
handleChange(event) {
  this.setState({value: event.target.value.toUpperCase()});
}
```

## Tag textarea {#the-textarea-tag}

Em HTML, o texto de um elemento <textarea> é definido usando seus filhos:

```html
<textarea>
  Apenas algum texto em uma área de texto
</textarea>
```

> Em React, o `<textarea>` usa um atributo `value`. Desta forma, um formulário usando um `<textarea>` pode ser escrito de forma muito semelhante a um formulário que usa um input de linha única:

Reviewer 2: ```suggestion
Em React, em vez disso, o `<textarea>` usa um atributo `value`. Desta forma, um formulário usando um `<textarea>` pode ser escrito de forma muito semelhante a um formulário que usa um input de linha única:
```

  }

  handleSubmit(event) {
    alert('Uma dissertação foi enviada: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Dissertação:
          <textarea value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Enviar" />
      </form>
    );
  }
}
```

Observe que `this.state.value` é inicializado no construtor, para que o textarea comece com algum texto.

## Tag select {#the-select-tag}

> Em HTML, `<select>` cria uma lista suspensa. Por exemplo, esse HTML cria uma lista suspensa de sabores:

Reviewer 2: ```suggestion
Em HTML, `<select>` cria uma lista suspensa (drop-down). Por exemplo, esse HTML cria uma lista suspensa de sabores:
```

```html
<select>
  <option value="laranja">Laranja</option>
  <option value="limão">Limão</option>
  <option selected value="coco">Coco</option>

>   <option value="manga">Manga</option>

Reviewer: I don't think the examples should be translated

Reviewer 2: @cstpk 

Nesse caso são só strings, você acha que prejudica muito?

Eu pensei em pedir pra ele reverter, mas a primeira sugestão do [styleguide](https://github.com/reactjs/pt-BR.reactjs.org/blob/master/GLOSSARY.md) é:

> Leave code blocks untranslated -- It is usually OK to translate comments and strings inside them.

O que acha?

Author: Eu não acho que prejudica, pelo contrário, ajuda no entendimento dos exemplos para pessoas que não sabem inglês.

Reviewer 3: @httpiago acho q pode manter assim, ou talvez deixar o conteúdo da option em português, mas o do atributo `value` deixaria em inglês, por exemplo, quando estou programando meu backend sempre é todo em inglês, apenas o front trada outros idiomas. Não sei se ajudei :see_no_evil:

Reviewer 4: ```suggestion
 <option value="laranja">Laranja</option>
  <option value="limao">Limão</option>
  <option selected value="coco">Coco</option>
  <option value="manga">Manga</option>
```

Reviewer 4: @jhonmike precisa estar os values em português, porém sem acentuação.

Reviewer 2: @glaucia86 @jhonmike @httpiago @cstpk 
Oi Pessoal, nesse caso em específico eu estou achando essa tradução das strings complicada. 

O que acordamos em relação a strings seria que "de forma geral", poderiam ser traduzidas, mas utilizando o bom senso para julgar onde não devem. Nesse caso aqui eu acho que adicionamos uma carga cognitiva muito grande em cima do programador pois criamos muitas possibilidades de confusão:

1.  Pode traduzir, mas no value não pode ter acento.
2. Se o value for comparado mas à frente com qualquer string, essa string vai ter que ser traduzida também, caso contrário estaríamos introduzindo um erro de lógica/comparação
3. Se qqr um dos valores não for traduzido, por acidente, estaríamos também introduzindo um erro, pois se o usuário copiar e colar o código e tiver uma string sem tradução, dependendo de como ela estiver sendo utilizada, poderá quebrar o código.
4. Se houver reutilização do código em qualquer outra parte da documentação, há a possibilidade dele não ter sido traduzido, e confundir o usuário.

Resumindo, eu acho que strings "inocentes" tipo labels ou alguma coisa mais simples que só faça parte da interface são ok para traduzir, mas é arriscado traduzir quando a string faz parte da lógica, cria-se muitos pontos frágeis para introdução de erros.

Enfim, essa é minha visão. Vejam aí se faz sentido!

</select>
```

> Note que a opção "coco" é selecionada por padrão, por causa do atributo `selected`. Em React, em vez de usar este atributo `selected`, usa-se um atributo `value` na raiz da tag `select`. Isso é mais conveniente em um componente controlado, porque você só precisa atualizá-lo em um só lugar. Por exemplo:

Reviewer: ```suggestion
Note que a opção Coconut é selecionada por padrão, por causa do atributo `selected`. Em React, em vez de usar este atributo `selected`, usa-se um atributo `value` na raiz da tag `select`. Isso é mais conveniente em um componente controlado, porque você só precisa atualizá-lo em um só lugar. Por exemplo:
```

Reviewer 4: @cstpk definimos no Glossário que as strings serão traduzidas. Sem problemas! ;)

```javascript{4,10-12,24}
class FlavorForm extends React.Component {
  constructor(props) {
    super(props);

>     this.state = {value: 'coco'};

Reviewer: ```suggestion
    this.state = {value: 'coconut'};
```

Reviewer 4: @cstpk novamente, no glossário os membros do React Community em Português decidiram que as strings devem ser traduzidas. Sem problemas o que foi feito aqui ;)

  }

  handleSubmit(event) {
    alert('Seu sabor favorito é: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Escolha seu sabor favorito:
          <select value={this.state.value} onChange={this.handleChange}>
            <option value="laranja">Laranja</option>

>             <option value="limão">Limão</option>

Reviewer 4: ```suggestion
<option value="limao">Limão</option>
```

            <option value="coco">Coco</option>
            <option value="manga">Manga</option>
          </select>
        </label>
        <input type="submit" value="Enviar" />
      </form>
    );
  }
}
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/JbbEzX?editors=0010)

No geral, isso faz com que as tags `<input type="text">`, `<textarea>` e `<select>` funcionem de forma muito semelhante - todos eles aceitam um atributo `value` que você pode usar para implementar um componente controlado.

> > Observação

Reviewer 2: ```suggestion
> Nota
```

>
> Você pode passar um array para o atributo `value`, permitindo que você selecione várias opções em uma tag `select`:
>
>```js
><select multiple={true} value={['B', 'C']}>
>```

> ## Tag de entrada de arquivo {#the-file-input-tag}

Reviewer 2: ```suggestion
## Tag de entrada de arquivo (file input) {#the-file-input-tag}
```

Em HTML, o `<input type="file">` permite ao usuário escolher um ou mais arquivos de seu dispositivo para serem enviados para um servidor ou manipulados por JavaScript através da [File API](https://developer.mozilla.org/pt-BR/docs/Web/API/File/Using_files_from_web_applications).

```html
<input type="file" />
```

> Como seu valor é de somente leitura, ele é um componente **não controlado** do React. É discutido melhor com outros componentes não controlados [mais adiante na documentação](/docs/uncontrolled-components.html#the-file-input-tag).

Reviewer 2: ```suggestion
Como seu valor é de somente leitura, ele é um componente **não controlado** do React. Esses são discutidos junto a outros componentes não controlados [mais adiante na documentação](/docs/uncontrolled-components.html#the-file-input-tag).
```

    return (
      <form>
        <label>

>           Estão indo:

Reviewer: ```suggestion
          Is going:
```

        </label>
        <br />
        <label>

>           Número de convidados:

Reviewer: ```suggestion
          Number of guests:
```

}
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/wgedvV?editors=0010)

Observe como usamos a sintaxe ES6 [nomes de propriedades computados] (https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Operators/Inicializador_Objeto#Nomes_de_propriedades_computados) para atualizar a chave de estado correspondente ao nome de entrada fornecido:

```js{2}
this.setState({
  [name]: value
});
```

É equivalente a este código no ES5:

```js{2}
var partialState = {};
partialState[name] = value;
this.setState(partialState);
```

Além disso, como o `setState()` automaticamente [mescla um estado parcial ao estado atual](/docs/state-and-lifecycle.html#state-updates-are-merged), nós podemos chamá-lo apenas com as partes alteradas.

> ## Valor Nulo em um input controlado {#controlled-input-null-value}

Reviewer 2: ```suggestion
## Valor Nulo em um Input Controlado {#controlled-input-null-value}
```
